### DISCUSSION TITLE
After using Temp Rand Provider, how to get OpenSSL to return to CTR-DRBG default rand provider?

### DISCUSSION THREAD

**Discussion by jmcpheters** [openssl/openssl#24870] [Q&A] [answered]
I am using OpenSSL 3.0.11 in a C program and I am trying to replace deprecated functions. Currently the program uses RAND_set_rand_method() for a proprietary random number generator for short while and then uses RAND_set_rand_method(RAND_OpenSSL()) for the remainder of program execution.

I created a simple "builtin" provider based on available sample code, which works ok. The problem is I cannot figure out how to get OpenSSL to go back to using the default provider CTR-DRBG at a later point.

I have a simplified code excerpt that illustrates the issue. For the provider I used the following source as the base: openssl-3.0.11/fuzz/fuzz_rand.c

For the code excerpt, I made the following changes to convert fuzz_rand.c to temp_rand.c:
1. removed header include not needed for my program: #include "fuzzer.h",
2. replaced all occurrences of "fuzz" to "temp" (e.g. fuzz_rand_generate -> temp_rand_generate),
3. renamed FuzzerSetRand() and FuzzerClearRand() to TempSetRand() and TempClearRand(),
4. added a printf to temp_rand_generate() 

The temp_rand_generate() function is as follows:

```
static int temp_rand_generate(ossl_unused void *vdrbg,
                              unsigned char *out, size_t outlen,
                              ossl_unused unsigned int strength,
                              ossl_unused int prediction_resistance,
                              ossl_unused const unsigned char *adin,
                              ossl_unused size_t adinlen)
{
    unsigned char val = 1;
    size_t i;

    printf("temp_rand_generate: called!\n");
    for (i = 0; i < outlen; i++)
        out[i] = val++;
    return 1;
}
```

I tested on Ubuntu 20.04, which comes with Openssl 1.1.1, so OpenSSL 3.0.11 was installed in an alternative location via:

```
$ ./Configure --prefix=/home/pmlinux64/local --openssldir=/home/pmlinux64/local/ssl -d
make
make install
```

The main function of the code excerpt (two_rands.c) is shown below, with the bare minimum needed to show the issue (so no error checking, initializations are simplified, etc).

```
#include <openssl/ssl.h>
#include <openssl/evp.h>
#include <openssl/rand.h>

void TempSetRand(void);         // from FuzzerSetRand()
void TempClearRand(void);       // from FuzzerClearRand() 

int main(int argc, char *pArgv[])
{
    EVP_PKEY_CTX  *pctx = NULL;
    EVP_PKEY          *key = NULL;
    SSL_METHOD    *ssl_method = NULL;
    SSL_CTX             *ctx = NULL;

    OpenSSL_add_all_algorithms();
    SSL_library_init();

    TempSetRand();
    // the following was done in TempSetRand:
    // OSSL_PROVIDER_add_builtin(NULL, "temp-rand", temp_rand_provider_init)
    // RAND_set_DRBG_type(NULL, "temp", NULL, NULL, NULL)
    // r_prov = OSSL_PROVIDER_try_load(NULL, "temp-rand", 1)

    // generate key
    pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_ED25519, NULL);
    EVP_PKEY_keygen_init(pctx);
    printf("main: generating key, temp_rand_generate wanted\n");
    EVP_PKEY_keygen(pctx, &key); // hits printf in temp_rand_generate (formally fuzz_rand_generate)

    // key used for initialization sequence, when done go back to the default DRGB

    TempClearRand();
    // the following was done in TempClearRand:
    // OSSL_PROVIDER_unload(r_prov)

    RAND_set_DRBG_type(NULL, "CTR-DRBG", NULL, NULL, NULL);

    // setup client
    ssl_method = (SSL_METHOD *)TLS_client_method();
    printf("Main: getting SSL context, temp_rand_generate NOT wanted\n");
    ctx = SSL_CTX_new(ssl_method); // hits printf in temp_rand_generate, why????
    // Why was temp_rand_generate hit when TempClearRand unloaded the provider, 
    // and RAND_set_DRGB_type was called with CTR-DRGB?

    return 0;
}
```

The code excerpt was compiled via:

```
gcc -I /home/pmlinux64/local/include -g -O0 -c two_rands.c -o two_rands.o
gcc -I /home/pmlinux64/local/include -g -O0 -c temp_rand.c -o temp_rand.o
gcc two_rands.o temp_rand.o /home/pmlinux64/local/lib64/libssl.a \ 
    /home/pmlinux64/local/lib64/libcrypto.a -o two_rands -lpthread -ldl
```

When executed, the following output is generated:

```
$ ./two_rands
main: generating key, temp_rand_generate wanted
temp_rand_generate: called!
main: getting SSL context, temp_rand_generate NOT wanted
temp_rand_generate: called!
```

Note that temp_rand_generate should only be called once, but is called twice because something is not correct with trying to get OpenSSL to go back to using CTR-DRGB. How can this be achieved? What is missing from the above code excerpt?

**Comment by paulidale** [comment]
You should use `RAND_set0_private()` and/or `RAND_set0_public()` to change the DRBG instances.  `RAND_set_DRBG_type()` only changes what gets loaded initially.  I.e. it only has an impact once and then only if it is called before the DRBGs are created.

**Reply by jmcpheters** [reply]
OpenSSL 3.0.11 is being used. When I add the new RAND function calls to two_rands.c and build, the link phase fails with:

```
undefined reference to `RAND_set0_private
undefined reference to `RAND_set0_public
```

When I look at the corresponding man pages at openssl.org, it only has links for 3.2 and 3.3. Does that mean it's not there for 3.0.11?

The plan is to go to the most recent version, so OpenSSL 3.3.1. However I was hoping to get rid of the deprecation warnings for 3.0.11 ASAP since the expectation is that future OpenSSL releases may turn the deprecations warnings to errors precluding successful builds.

So is there a way to go back to CTR-DBRG that is not too onerous for version 3.0.11? I'd like to keep the code as simply as possible, so if getting it to work in 3.0.11 is complicated, I will finish eliminating other deprecation warnings and I should be able to get to OpenSSL 3.3.1 in a week or two.

I do have a question about the man page and it's references to threads for RAND_set0_private/public(). The code excerpt, two_rands.c, initializes the temporary rand with no threads spawned, so just in the main thread. When the program is done with the temp rand, it needs to reset to the default rand (CTL-DBRG), also in the main thread. There will be SSL_CTX_new() calls in subsequently spawned threads. Does that mean each individual thread needs to call RAND_set0_private/public() too? Normally I would just try with some test code, but since an upgrade to OpenSSL 3.3.1 will not be done right away, I thought I'd ask in case someone knows the answer off the top of their head.

**Reply by jmcpheters** [reply]
The RAND_set0_public/private() man page start starts at 3.1 (not 3.2 that I previously stated). However, it still will not build for Openssl 3.0.11, and when I search through rand.h I can't find the setters. Curiously the getters are there. See below...

```
$ grep set0 /home/pmlinux64/local/include/openssl/rand.h
$ grep get0 /home/pmlinux64/local/include/openssl/rand.h
EVP_RAND_CTX *RAND_get0_primary(OSSL_LIB_CTX *ctx);
EVP_RAND_CTX *RAND_get0_public(OSSL_LIB_CTX *ctx);
EVP_RAND_CTX *RAND_get0_private(OSSL_LIB_CTX *ctx);
$
```

**Reply by paulidale** [reply]
In 3.0 this becomes difficult.  The only way I can think of doing it is to create a new library context for this.  Create and use the test-rand's inside this context only.

**Reply by jmcpheters** [reply]
I unmarked as answered because I still can't get it working with OpenSSL 3.3.1, and I will momentarily post more details on the next problem encountered.

**Reply by jmcpheters** [reply]
I updated by to OpenSSL 3.3.1, over writing the installation for OpenSSL 3.0.11. It now builds because Rand_set0_private/public() exist. I updated the TempRandClear() to be the following:

```
void TempClearRand(void)
{
    EVP_RAND     *rand = NULL;
    EVP_RAND_CTX *rctx = NULL;

    OSSL_PROVIDER_unload(r_prov);
    rand = EVP_RAND_fetch(NULL, "CTR-DRBG", NULL);
    rctx = EVP_RAND_CTX_new(rand, NULL);
    EVP_RAND_free(rand);
    RAND_set0_private(NULL, rctx);
    RAND_set0_public(NULL, rctx);
    EVP_RAND_CTX_free(rctx);
}
```

In main(), I commented out the call to RAND_set_DRBG_type(NULL, "CTR-DRBG", NULL, NULL, NULL) as it didn't appear to do anything (a quick check inside said it was already instantiated).

After making the calls to RAND_set0_private and RAND_set0_public, the printf in temp_rand_generate() was never hit, which is the desired behavior because the default CTR-DRBG is wanted, not the temp_rand provider.

Unfortunately, when main() calls SSL_CTX_new(ssl_method), it SegFaults in evp_rand_lock() (from evp_rand.c) at the line with the check:

```
    if (rand->->lock != NULL)
```

The stack back trace is:

```
evp_rand_lock(EVP_RAND_CTX * rand) (\pm\g7\online_utilities\external\openssl-3.3.1\openssl-3.3.1\crypto\evp\evp_rand.c:105)
EVP_RAND_generate(EVP_RAND_CTX * ctx, unsigned char * out, size_t outlen, unsigned int strength, int prediction_resistance, const unsigned char * addin, size_t addin_len) (\pm\g7\online_utilities\external\openssl-3.3.1\openssl-3.3.1\crypto\evp\evp_rand.c:590)
RAND_bytes_ex(OSSL_LIB_CTX * ctx, unsigned char * buf, size_t num, unsigned int strength) (\pm\g7\online_utilities\external\openssl-3.3.1\openssl-3.3.1\crypto\rand\rand_lib.c:380)
SSL_CTX_new_ex(OSSL_LIB_CTX * libctx, const char * propq, const SSL_METHOD * meth) (\pm\g7\online_utilities\external\openssl-3.3.1\openssl-3.3.1\ssl\ssl_lib.c:3998)
SSL_CTX_new(const SSL_METHOD * meth) (\pm\g7\online_utilities\external\openssl-3.3.1\openssl-3.3.1\ssl\ssl_lib.c:4104)
main(int argc, char ** pArgv) (\pm\g7\online_utilities\Excerpt\two_rands.c:50)
```

A screenshot showing the segfault in all it's glory:

![Rand_set0_linux](https://github.com/user-attachments/assets/0a3bf3ea-c4b5-4a8e-a1a6-92cf076a035c)

When I comment out the calls to TempSetRand() and TempClearRand(), so that my temp_rand provider is not used,  the SegFault goes away. I also tried removing the EVP_RAND_free() and EVP_RAND_CTX_free() calls in TempClearRand(), thinking maybe by leaving little memory leak that the segfault will not happen. However a different segfault occurred, which you can see in the next screenshot:

![Rand_set0_linux_no_free](https://github.com/user-attachments/assets/582911bc-8b79-4076-b865-24bf28ce0a27)

I tried the same tests for a Windows build with similar results. When the temp_rand provider is not instantiated and used, there is no segfault. When rand and rctx are not freed in TempClearRand(), the segfault for Windows is in the same location as the segfault for Linux. However, when rand and rctx are freed in TempClearRand(), the segfault for Windows is in a separate location, as shown in the screenshot:

![Rand_set0_windows](https://github.com/user-attachments/assets/69dbdea0-69cb-4963-8b6a-8860794cbe66)

At this point, my assumption is the Rand_set0_private/public() use is not correct, but I'm at a loss of what is wrong.

**Reply by jmcpheters** [reply]
Sorry for typo on my line for the first segfault example, should read: if (rand->meth->lock != NULL)

**Comment by jmcpheters** [comment]
Maybe this is the answer?

I added a libctx to main() in two_rands.c:

```
    OSSL_LIB_CTX *libctx = OSSL_LIB_CTX_new();
```

Then I changed the call to EVP_PKEY_CTX_new_id() in main() to be the following:

```
    pctx = EVP_PKEY_CTX_new_from_name(libctx, "ED25519", NULL);
```

Then I redefined TempSetRand() and TempClearRand() to use the new libctx as follows:

```
void TempSetRand(OSSL_LIB_CTX *lctx)
{
    if (!OSSL_PROVIDER_add_builtin(lctx, "temp-rand", temp_rand_provider_init)
        || !RAND_set_DRBG_type(lctx, "temp", NULL, NULL, NULL)
        || (r_prov = OSSL_PROVIDER_try_load(lctx, "temp-rand", 1)) == NULL)
        exit(1);
}

void TempClearRand(OSSL_LIB_CTX *lctx)
{
    OSSL_PROVIDER_unload(r_prov);
    OSSL_LIB_CTX_free(lctx);
}
```

After updating the  TempSetRand() and TempClearRand() calls in main() to pass in libctx, expected behavior was achieved, namely:

1. temp_rand_generate() gets called when generating the key via EVP_PKEY_keygen().
2. temp_rand_generate() does NOT get called when calling SSL_CTX_new().
3. No SegFault happens.

Are the changes I outlined above a valid implementation? 
Should it have worked the other way where a new libctx was not created, but everything is under the default libctx (NULL)?

**Reply by paulidale** [reply, marked as answer]
Yes, this is what I meant [earlier](https://github.com/openssl/openssl/discussions/24870#discussioncomment-10067134).

**Reply by jmcpheters** [reply]
Got it. But the solution was not very complicated, and Rand_set0_public/private() was not used in the end, so I was a slightly bewildered...
